File: szavazas-core/src/main/java/hu/kdea/szavazas/ballotprocessor/grid/GridDetectorStep.java
package hu.kdea.szavazas.ballotprocessor.grid;

import boofcv.struct.image.GrayU8;
import hu.kdea.szavazas.ballotprocessor.common.Rect;
import java.util.ArrayList;
import java.util.List;
import javax.inject.Inject;

public class GridDetectorStep {
    private final GridDetectionOrchestrator orchestrator;

    @Inject
    public GridDetectorStep(GridDetectionOrchestrator orchestrator) {
        this.orchestrator = orchestrator;
    }

    public List<Rect> detect(GrayU8 projectionInput, int cropTop, int qrCentreX, int expectedCols, int expectedRows) {
        Rect rect = new Rect(0, 0, projectionInput.width, projectionInput.height);
        List<Rect> boxes = orchestrator.detect(projectionInput, rect, expectedCols, expectedRows, true, true);
        if (boxes.isEmpty()) {
            return null;
        }
        List<Rect> translated = new ArrayList<>();
        for (Rect box : boxes) {
            translated.add(new Rect(box.getX() + qrCentreX, box.getY() + cropTop, box.getWidth(), box.getHeight()));
        }
        return translated;
    }
}

File: szavazas-core/src/main/java/hu/kdea/szavazas/ballotprocessor/grid/GridDetectorStep.kt
package hu.kdea.szavazas.ballotprocessor.grid

import boofcv.struct.image.GrayU8
import hu.kdea.szavazas.ballotprocessor.common.Rect

class GridDetectorStep(private val orchestrator: GridDetectionOrchestrator) {
    fun detect(
        projectionInput: GrayU8,
        cropTop: Int,
        qrCentreX: Int,
        expectedCols: Int,
        expectedRows: Int
    ): List<Rect>? {
        val rect = Rect(0, 0, projectionInput.width, projectionInput.height)
        val boxes = orchestrator.detect(
            binaryClosed = projectionInput,
            searchRect = rect,
            expectedCols = expectedCols,
            expectedRows = expectedRows,
            skipBoundaries = true,
            emptySecondColumn = true
        )
        if (boxes.isEmpty()) return null
        return boxes.map { Rect(it.x + qrCentreX, it.y + cropTop, it.width, it.height) }
    }
}